### ISSUE TITLE
Lack of Input Validation in theano.gradient.verify_grad

### ISSUE THREAD

**Issue by digitalali97** [Theano/Theano#6813]
Issue with theano.gradient.verify_grad Failing to Check Input Validity
The theano.gradient.verify_grad function does not validate the compatibility of input shapes for the tested function. In the example below, input_1 and input_alpha are of mismatched shapes, but theano.gradient.verify_grad fails to detect the issue and outputs None without any error. However, directly using these inputs with theano.tensor.nnet.relu causes an error.

**Problematic Code**
`import theano
import theano.tensor as T
import numpy as np

# Custom activation function using Theano's ReLU
def custom_activation(x, alpha):
    return theano.tensor.nnet.relu(x, alpha)

# Define symbolic tensors
x1 = T.tensor3('x1')
alpha = T.tensor4('alpha')

# Compute output and gradients
output = custom_activation(x1, alpha)
loss = T.sum(output ** 2)
grad_x1, grad_alpha = T.grad(loss, [x1, alpha])

# Define inputs with mismatched shapes
input_1 = np.random.random((1, 2, 3)).astype('float32')       # Shape: (1, 2, 3)
input_alpha = np.random.random((1, 2, 1, 1)).astype('float32')  # Shape: (1, 2, 1, 1)

# Verify gradients
rng = np.random.RandomState(123)
print(theano.gradient.verify_grad(custom_activation, pt=[input_1, input_alpha], rng=rng))
`

Explanation of the **Issue**
Shape Mismatch: The inputs input_1 and input_alpha must have compatible shapes for broadcasting when passed to theano.tensor.nnet.relu. In this case, they are not compatible.
No Error from verify_grad: Despite the incompatibility, theano.gradient.verify_grad does not validate the input shapes before proceeding, leading to an incorrect None output instead of raising an error.
Error in theano.tensor.nnet.relu: If custom_activation is used directly, the underlying theano.tensor.nnet.relu function correctly raises an error due to the shape mismatch.

**Comment by micedevai** [comment]
The issue you are encountering with `theano.gradient.verify_grad` failing to validate input shape compatibility is rooted in the fact that `verify_grad` does not check for shape compatibility before evaluating gradients. This lack of input validation can result in the `verify_grad` function proceeding without raising an error, which leads to an incorrect result (`None`), even when there's an issue with the shape of the inputs.

Here's a more detailed breakdown of the problem and possible solutions:

### Explanation of the Issue

1. **Shape Mismatch:**
   - In your example, `input_1` has the shape `(1, 2, 3)` and `input_alpha` has the shape `(1, 2, 1, 1)`.
   - When these inputs are passed to `theano.tensor.nnet.relu`, Theano tries to perform an operation involving these tensors. However, `input_1` and `input_alpha` need to have compatible shapes for broadcasting to work correctly. The shapes `(1, 2, 3)` and `(1, 2, 1, 1)` are **not broadcast-compatible**, which causes a shape mismatch.
   
2. **No Error from `verify_grad`:**
   - `theano.gradient.verify_grad` tests the gradients of the function with respect to the inputs, but it does not perform any checks on the compatibility of the input shapes for broadcasting or for the operation itself. Therefore, it proceeds with the mismatch and returns `None`, which is not informative about the actual problem.
   
3. **Direct Usage with `relu`:**
   - When you directly pass the mismatched shapes `(1, 2, 3)` and `(1, 2, 1, 1)` to `theano.tensor.nnet.relu`, Theano correctly raises a **broadcasting error** because the shapes are not compatible.

### Solution 1: Manually Check Shape Compatibility

Before running `verify_grad`, you can manually check whether the input shapes are compatible for broadcasting. This can be done by using `theano.tensor.broadcast` or a simple shape check, like this:

```python
import theano
import theano.tensor as T
import numpy as np

# Custom activation function using Theano's ReLU
def custom_activation(x, alpha):
    return theano.tensor.nnet.relu(x, alpha)

# Define symbolic tensors
x1 = T.tensor3('x1')
alpha = T.tensor4('alpha')

# Compute output and gradients
output = custom_activation(x1, alpha)
loss = T.sum(output ** 2)
grad_x1, grad_alpha = T.grad(loss, [x1, alpha])

# Define inputs with mismatched shapes
input_1 = np.random.random((1, 2, 3)).astype('float32')  # Shape: (1, 2, 3)
input_alpha = np.random.random((1, 2, 1, 1)).astype('float32')  # Shape: (1, 2, 1, 1)

# Check if the shapes are compatible for broadcasting
try:
    # Try to perform a broadcast operation
    T.broadcast(input_1, input_alpha)
    print("Shapes are compatible.")
except Exception as e:
    print(f"Shape mismatch error: {e}")

# Verify gradients only if the shapes are compatible
if input_1.shape == input_alpha.shape:
    rng = np.random.RandomState(123)
    print(theano.gradient.verify_grad(custom_activation, pt=[input_1, input_alpha], rng=rng))
else:
    print("Input shapes are not compatible for broadcasting.")
```

### Solution 2: Handle Shape Broadcasting in `custom_activation`

If you're trying to use broadcasting correctly, you should ensure that the shapes are compatible in your custom activation function. You can do this explicitly before calling `relu` by reshaping or modifying the tensors to make them broadcast-compatible. Here’s an example:

```python
import theano
import theano.tensor as T
import numpy as np

# Custom activation function using Theano's ReLU
def custom_activation(x, alpha):
    # Ensure the shapes are broadcast-compatible
    # Reshape alpha to match the shape of x if needed
    alpha_reshaped = alpha.dimshuffle(0, 1, 'x', 'x')  # Reshape alpha to (1, 2, 1, 1)
    return theano.tensor.nnet.relu(x, alpha_reshaped)

# Define symbolic tensors
x1 = T.tensor3('x1')
alpha = T.tensor4('alpha')

# Compute output and gradients
output = custom_activation(x1, alpha)
loss = T.sum(output ** 2)
grad_x1, grad_alpha = T.grad(loss, [x1, alpha])

# Define inputs with mismatched shapes
input_1 = np.random.random((1, 2, 3)).astype('float32')  # Shape: (1, 2, 3)
input_alpha = np.random.random((1, 2, 1, 1)).astype('float32')  # Shape: (1, 2, 1, 1)

# Verify gradients
rng = np.random.RandomState(123)
print(theano.gradient.verify_grad(custom_activation, pt=[input_1, input_alpha], rng=rng))
```

### Solution 3: Use `theano.gradient.verify_grad` with Correct Input Shapes

Ensure that the inputs you're passing to `verify_grad` are broadcast-compatible. You can modify the shape of `input_alpha` or `input_1` to make them compatible for broadcasting, for example:

```python
# Ensure input_1 and input_alpha are compatible for broadcasting
input_1 = np.random.random((1, 2, 3)).astype('float32')  # Shape: (1, 2, 3)
input_alpha = np.random.random((1, 2, 3)).astype('float32')  # Modify shape to match input_1

# Verify gradients with compatible shapes
rng = np.random.RandomState(123)
print(theano.gradient.verify_grad(custom_activation, pt=[input_1, input_alpha], rng=rng))
```

### Solution 4: Use `verify_grad` with Known Compatible Shapes

If you want to stick with using `verify_grad`, ensure that the input shapes you're testing are **explicitly compatible** for the broadcasting rules of the operations you're using. The example below ensures the shapes match correctly for broadcasting:

```python
# Modify input_alpha to have the same shape as input_1
input_alpha = np.random.random((1, 2, 3)).astype('float32')

# Verify gradients
rng = np.random.RandomState(123)
print(theano.gradient.verify_grad(custom_activation, pt=[input_1, input_alpha], rng=rng))
```

The root cause of the issue is that `theano.gradient.verify_grad` doesn't check for broadcasting compatibility or shape mismatches, and it proceeds with the inputs even when they're not compatible. To fix the issue, you can either manually check the shapes before running `verify_grad`, reshape the inputs to ensure they are compatible for broadcasting, or modify the inputs so they are compatible.
